src/app/reporting/Engine.cpp

```diff
@@ -1009,8 +1009,10 @@ void Engine::MarkDirty(const ConcreteAttributePath & path)
 {
     // NOTE: original MatterReportingAttributeChangeCallback ALSO does a
     //       IncreaseClusterDataVersion. However that ties directly to ember so it is not
-    //       done here. The code should be migrated to execute this.
-
+    //       done here.
+    //
+    //       Instead the DataModel::Provider is responsible to properly increase cluster
+    //       data versions.
```

Comment:
Fundamentally: any time things are marked as dirty for reporting that _must_ also increase cluster data versions.  Anything that decouples those things is going to be a source of bugs...

So if the intent is that the DataModel::Provider both bumps data versions and calls `Engine::MarkDirty`, that's fine; the ember version can do that via its current function it calls.

Reply:
Existing solution is circular dependency-wise: ember calls reporting changed (mark dirty) and reporting needs to change version ... which is held by ember.

I could add versioning calls to the Provider interface for specifically notifying of version changes, however this seems odd: Provider already handles versions (they are returned as part of iteration since we can filter on them) and provider calls `MarkDirty`, so it seems odd for a provider to say `mark dirty` just to get back a callback of `also increase version` which we already knew it is needed when we called MarkDirty in the first place.